File: 1047-remove-all-adjacent-duplicates-in-string/1047-remove-all-adjacent-duplicates-in-string.cpp
class Solution {
public:
    string removeDuplicates(string s) {
            
            stack<char>st;
            for(int i=0;i<s.size();i++){
                    if(st.empty())
                            st.push(s[i]);
                    else if(st.top()==s[i])
                            st.pop();
                    else st.push(s[i]);
            }
            string a="";
            while(!st.empty()){
                    a+=st.top();
                    st.pop();
            }
            reverse(a.begin(),a.end());
            return a;
    }
};
            
            /*
            without stack optimal
         string ans="";
        for(char i:s){
            if(ans.size()==0)ans.push_back(i);//if empty just push each elements of string s
            else if(i==ans.back())ans.pop_back();//if mathed with next element remove last element pushed in ans string
            else ans.push_back(i);//else just push the element simply
        }
        return ans;
    }
};*/

File: 1047-remove-all-adjacent-duplicates-in-string/1047-remove-all-adjacent-duplicates-in-string.java
class Solution {
    public String removeDuplicates(String s) {
//method 2 Stack
Stack<Character>st=new Stack<>();
        for(int i=0;i<s.length();i++){
            if(!st.empty() && s.charAt(i)==st.peek())
                st.pop();
            else 
                st.add(s.charAt(i));                
        }
        StringBuilder sb=new StringBuilder();
        while(!st.empty())
        sb.append(st.pop());
        return sb.reverse().toString();
    }
}
    /*
//method 1 StringBuilder
       StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (sb.length() > 0 && sb.charAt(sb.length() - 1) == c) {
                System.out.println("if");
                sb.deleteCharAt(sb.length() - 1);
            } else {
                System.out.println("else");
                sb.append(c);
            }
        }
        return sb.toString();
        }
        }
        */